File: python/matel.py
"""
Matrix-element computation for NECG.

Analogous to matel.f90.  Computes the overlap (ov), kinetic-energy (tk),
electron-repulsion (ert), and electric-field (zht) contributions for a
pair of Correlated-Gaussian basis functions ii and jj under symmetry
term stf.

All indices are **1-based** on input (matching Fortran call sites) and
converted internally.

Key NumPy replacements:
  - Fortran ``matmul``        → ``@`` operator
  - Fortran DPOTRF + DPOTRI   → ``numpy.linalg.inv``
  - Kronecker products        → ``numpy.kron(A, np.eye(3))``
"""

from __future__ import annotations

import math

import numpy as np

import shared as sh
from utils import f0


def matel(stf: int, ii: int, jj: int) -> tuple[float, float, float, float]:
    """
    Compute matrix elements between basis functions *ii* and *jj* under
    symmetry term *stf* (all 1-based, matching Fortran convention).

    Returns
    -------
    ov  : overlap integral
    tk  : kinetic-energy integral
    ert : electron-repulsion integral (summed over all pairs)
    zht : electric-field (Stark) integral
    """
    # Convert to 0-based indices
    stf0 = stf - 1
    ii0 = ii - 1
    jj0 = jj - 1

    ne = sh.ne
    svh = sh.svh
    skp = sh.skp
    blk = svh + skp   # size of one basis-function block in iv

    # ------------------------------------------------------------------
    # Build lower-triangular Cholesky factors lk (bra) and ell (ket)
    # from the iv vector.
    # iv[(ii0*blk) .. (ii0*blk + svh - 1)] holds vech(lk)
    # iv[(jj0*blk) .. (jj0*blk + svh - 1)] holds vech(ell)
    # ------------------------------------------------------------------
    lk = np.zeros((ne, ne))
    ell = np.zeros((ne, ne))

    kk = ii0 * blk
    ll = jj0 * blk
    for i in range(ne):          # column (Fortran i)
        for j in range(i, ne):   # row   (Fortran j >= i → lower triangle)
            lk[j, i] = sh.iv[kk]
            ell[j, i] = sh.iv[ll]
            kk += 1
            ll += 1

    # Positive-semidefinite matrices  Ak = lk lk^T,  Al = ell ell^T
    ak = lk @ lk.T
    al = ell @ ell.T

    # ------------------------------------------------------------------
    # Apply symmetry operation on the ket:  al → sml^T · al · sml
    # Fortran: alt = al @ sml;  al = sml^T @ alt
    # ------------------------------------------------------------------
    sml_k = sh.sml[:, :, stf0]
    al = sml_k.T @ (al @ sml_k)

    # ------------------------------------------------------------------
    # Shift vectors  st1 (bra) and st2 (ket)
    # iv[(ii0*blk + svh) .. (ii0*blk + svh + skp - 1)]
    # ------------------------------------------------------------------
    st1 = sh.iv[ii0 * blk + svh: ii0 * blk + svh + skp].copy()
    st2 = sh.iv[jj0 * blk + svh: jj0 * blk + svh + skp].copy()

    # Apply symmetry on ket shift vector
    st2 = sh.sms[:, :, stf0] @ st2

    # ------------------------------------------------------------------
    # Matrix inverses via numpy (replacing DPOTRF + DPOTRI)
    # ------------------------------------------------------------------
    akl = ak + al
    akli = np.linalg.inv(akl)   # (ak + al)^{-1}

    # ------------------------------------------------------------------
    # Determinant factor.
    # Fortran: DPOTRF fills L (Cholesky of akl); dett = prod(L_ii)^2 = det(akl)
    # ------------------------------------------------------------------
    dett = abs(np.linalg.det(akl))   # = det(ak + al)

    # Product of diagonal elements of the Cholesky factors of ak and al
    # (det of lower-triangular lk / ell = product of diagonal)
    detak = math.sqrt(abs(np.prod(np.diag(lk)) ** 2))
    detal = math.sqrt(abs(np.prod(np.diag(ell)) ** 2))

    # ------------------------------------------------------------------
    # Kronecker products  (ne × ne → skp × skp block-diagonal matrices)
    # np.kron(A, eye(3)) produces the same result as the triple-k loop
    # in the Fortran source.
    # ------------------------------------------------------------------
    I3 = np.eye(3)
    akb = np.kron(ak, I3)
    alb = np.kron(al, I3)
    aklib = np.kron(akli, I3)
    massb = np.kron(sh.mass, I3)

    # ------------------------------------------------------------------
    # Overlap integral
    # i1 = st1^T akb st1,  i2 = st2^T alb st2
    # ee = akb·st1 + alb·st2
    # st = aklib · ee
    # i3 = ee^T · aklib · ee = ee^T · st
    # ov = exp(-(i1+i2)+i3) * sqrt(8^ne) * sqrt((detak·detal/dett)^3)
    # ------------------------------------------------------------------
    iv1 = akb @ st1
    i1 = float(st1 @ iv1)

    iv2 = alb @ st2
    i2 = float(st2 @ iv2)

    ee = iv1 + iv2
    st = aklib @ ee
    i3 = float(ee @ st)

    prefac = math.sqrt(8.0 ** ne) * math.sqrt((detak * detal / dett) ** 3)
    ov = math.exp(-(i1 + i2) + i3) * prefac

    # ------------------------------------------------------------------
    # Kinetic-energy integral
    # tk = 2·ov·(2·i3_ke + trace_ke)
    # i3_ke = (st - st1)^T · (akb · massb · alb) · (st - st2)
    # trace_ke = tr(massb · akb · aklib · alb)
    # ------------------------------------------------------------------
    im1 = massb @ alb
    im2 = akb @ im1
    dif1 = st - st1
    dif2 = st - st2
    ke_i3 = float(dif1 @ (im2 @ dif2))

    im1 = aklib @ alb
    im2 = akb @ im1
    im1 = massb @ im2
    ke_i2 = float(np.trace(im1))

    tk = 2.0 * ov * (2.0 * ke_i3 + ke_i2)

    # ------------------------------------------------------------------
    # Electron-repulsion and nuclear-attraction integrals
    # Loop over all pairs (pp, ll_) with 0 ≤ pp < ll_ ≤ ne.
    # pp = 0 → nucleus–electron;  pp ≥ 1 → electron–electron.
    # ------------------------------------------------------------------
    ert = 0.0
    # pre-compute ism1 and store for each pair
    for pp in range(ne + 1):
        for ll_ in range(pp + 1, ne + 1):
            # Build the J matrix (ne × ne) — Fortran 1-based indices
            jij = np.zeros((ne, ne))
            if pp == 0:
                # nucleus–electron: only diagonal element for particle ll_
                jij[ll_ - 1, ll_ - 1] = 1.0
            else:
                jij[pp - 1, pp - 1] = 1.0
                jij[ll_ - 1, ll_ - 1] = 1.0
                jij[pp - 1, ll_ - 1] = -1.0
                jij[ll_ - 1, pp - 1] = -1.0

            # ism1 = jij · akli,  jij ← akli · ism1 = akli · jij_orig · akli
            ism1 = jij @ akli
            jij_new = akli @ ism1

            # Kronecker product of updated jij
            jijb = np.kron(jij_new, I3)

            # i3_er = ee^T · jijb · ee
            iv3 = jijb @ ee
            i3_er = float(ee @ iv3)

            # i2_er = tr(ism1) = tr(jij_orig · akli)
            i2_er = float(np.trace(ism1))

            if i2_er == 0.0:
                continue

            i1_er = i3_er / i2_er
            er = (sh.ncv[pp] * sh.ncv[ll_]
                  * 2.0 * ov * f0(i1_er) / math.sqrt(i2_er * sh.PI))
            ert += er

    # ------------------------------------------------------------------
    # Electric-field (Stark) term
    # zht = ov · efs · sum_i( ncv[i] · st[3*(i-1)+2] )
    # where i runs over electrons (1-indexed), st[3*(i-1)+2] is the z-
    # component (index 3*i in Fortran 1-based = 3*(i-1)+2 in Python).
    # ------------------------------------------------------------------
    zht_sum = 0.0
    for i in range(1, ne + 1):          # Fortran i = 1..ne
        zht_sum += sh.ncv[i] * st[3 * i - 1]   # st(3*i) in Fortran 1-based
    zht = ov * zht_sum * sh.efs

    return ov, tk, ert, zht

File: source/matel.f90
SUBROUTINE matel(stf,ii,jj,OV,tk,ert,zht)
!**********************************
!4/28/00 -- ov, tk, na, er are great; 
!	    

USE matrices_to_be_shared
implicit none		

double precision::zht
double precision::ov,tk,er,na,i1,i2,i3,j1,j2,j3,detakl,detak,detal,dett,f0,df0,ert,nat
double precision, dimension(ne,ne)  :: ak,al,lk,lkt,ell,ellt,alt		  
double precision, dimension(ne,ne)  ::akl,akli,jij,ism1,jije,aki,ism2,ism3,jii
double precision, dimension(skp,2)  ::si
double precision, dimension(3,ne)   ::dot1,dot1e
double precision, dimension(ne,3)   ::dot2,dot2e
double precision, dimension(skp,skp)::akb,alb,aklib,im1,im2,im3,iden,jijb,aklp,aklpt,jiib,massb
double precision, dimension(skp)    ::iv1,iv2,ee,iv3,st,st1,st2,en,nv
double precision, dimension(svh)    ::vech1
integer::zq,I,J,k,ii,jj,kk,ll,pp,stf
!******** make ak, al, sk, sl **************************

lk=zero
lkt=zero
ell=zero
ellt=zero
kk=(ii-1)*(svh+skp)+1
ll=(jj-1)*(svh+skp)+1
do i=1,ne
   do j=i,ne
      lk(j,i)=iv(kk)
      ell(j,i)=iv(ll)
      kk=kk+1
      ll=ll+1
   end do
end do

do j=1,ne 
     do i=1,j
      lkt(i,j)=lk(j,i)
      ellt(i,j)=ell(j,i)
   end do
end do

ak=matmul(lk,lkt)
al=matmul(ell,ellt)
!call tab(lk,ne,ne,ne,ne)
!stop
!******* perform symmetry operation on the ket
alt=zero
do i=1,ne
   do j=1,ne
      do k=1,ne
         alt(i,j)=alt(i,j)+al(i,k)*sml(k,j,stf)
      end do
   end do
end do

al=zero
do i=1,ne
   do j=1,ne
      do k=1,ne
         al(i,j)=al(i,j)+sml(k,i,stf)*alt(k,j)
      end do
   end do
end do
!call tab(al,ne,ne,ne,ne)
!stop
!********************************************
!!$call tab(ak,ne,ne,ne,ne)
!call tab(al,ne,ne,ne,ne)

kk=(ii-1)*(svh+skp)+svh+1
ll=(jj-1)*(svh+skp)+svh+1
do i=1,skp
   si(i,1)=iv(kk)
   si(i,2)=iv(ll)
   kk=kk+1
   ll=ll+1
end do

!write(*,*)si(1,2), si(2,2),si(3,2), si(4,2),si(5,2), si(6,2),si(7,2), si(8,2),si(9,2)
!******* perform symmetry operation on the ket shift vector
iv1=zero
do i=1,skp
   do j=1,skp
      iv1(i)=iv1(i)+sms(i,j,stf)*si(j,2)
   end do
end do
do i=1,skp
   si(i,2)=iv1(i)
end do
!***********************************************************
!write(*,*)si(1,2), si(2,2),si(3,2), si(4,2),si(5,2), si(6,2),si(7,2), si(8,2),si(9,2)
!stop
!*** st's defined here 
do i=1,skp
   st1(i)=si(i,1)
   st2(i)=si(i,2)
end do



akb=zero
alb=zero
aklib=zero
massb=zero

!write(*,*)st1
!write(*,*)st2
!!$call tab(ak,2,2,2,2)
!!$call tab(al,2,2,2,2)

akl=ak+al
akli=akl
aki=ak
call DPOTRF('l',ne,akli,ne,jjj)
call DPOTRF('l',ne,aki,ne,jjj)
dett=one
do i=1,ne
	dett=dett*akli(i,i)
end do
dett=dett**two
call DPOTRI('l',ne,akli,ne,jjj)
call DPOTRI('l',ne,aki,ne,jjj)

do i=1,ne
	do j=(i+1),ne
	aki(i,j)=aki(j,i)
	akli(i,j)=akli(j,i)
	end do
end do

!****************************************
!do the kronecker product

do i=1,ne
   kk=(i-1)
   do j=1,ne
      zq=(j-1)
      do k=1,3
      akb(3*kk+k,3*zq+k)=ak(i,j)
      massb(3*kk+k,3*zq+k)=mass(i,j)
      alb(3*kk+k,3*zq+k)=al(i,j)
      aklib(3*kk+k,3*zq+k)=akli(i,j)
      end do
   end do
end do
!call tab(ak,3,3,3,3)
!call tab(akb,9,9,9,9)
!stop

!******************************************

detal=one   
detak=one
do i=1,ne
	detal=detal*ell(i,i)
	detak=detak*lk(i,i)
end do
detal=detal**two
detak=detak**two
   
detal=dsqrt(detal)
detak=dsqrt(detak)



!****************************************************
! make it agree with L&K



!****************************************************
!   Do the integrals
!
!overlap integral 
   
iv1=zero
i1=zero
do i=1,skp
   do j=1,skp
      iv1(i)=iv1(i)+akb(i,j)*st1(j)
   end do
end do
do i=1,skp
   i1=i1+st1(i)*iv1(i)
end do

iv2=zero
i2=zero
do i=1,skp
   do j=1,skp
      iv2(i)=iv2(i)+alb(i,j)*st2(j)
   end do
end do
do i=1,skp
   i2=i2+st2(i)*iv2(i)
end do

!******* ee is defined for here on out
ee=iv1+iv2

iv3=zero
i3=zero
do i=1,skp
   do j=1,skp
      iv3(i)=iv3(i)+aklib(i,j)*ee(j)
   end do
end do
do i=1,skp
   i3=i3+ee(i)*iv3(i)
end do

!write(*,*)i3
st=iv3

ov=exp(-(i1+i2)+i3)*dsqrt(eight**ne)*dsqrt((detak*detal/dett)**3)
!!$write(*,*)i1,i2,i3
!!$ov=exp(-(i1+i2)+i3)

!********************************************************
!Kinetic energy 

iv1=zero
iv2=zero
i1=zero
i2=zero
i3=zero


im1=matmul(massb,alb)
im2=matmul(akb,im1)

do i=1,skp
   do j=1,skp
      i3=i3+(st(i)-st1(i))*im2(i,j)*(st(j)-st2(j))
   end do
end do


!!$do i=1,skp
!!$   do j=1,skp
!!$      iv1(i)=iv1(i)+alb(i,j)*(st(j)-st2(j))
!!$   end do
!!$end do
!!$
!!$
!!$do i=1,skp
!!$   do j=1,skp
!!$      iv2(i)=iv2(i)+massb(i,j)*iv1(j)
!!$   end do
!!$end do
!!$
!!$do i=1,skp
!!$   do j=1,skp
!!$      iv1(i)=iv1(i)+akb(i,j)*iv2(j)
!!$   end do
!!$end do
!!$
!!$do i=1,skp
!!$   i3=i3+(st(i)-st1(i))*iv1(i)
!!$end do

im1=matmul(aklib,alb)
im2=matmul(akb,im1)
im1=matmul(massb,im2)
im2=im1

do i=1,skp
   i2=i2+im2(i,i)
end do


tk=two*ov*(two*i3+i2)


!********************************************
! electron replsion -- associated gradients

er=zero
ert=zero

do pp=0,ne
   do ll=pp,ne
      if (pp==ll)then
         er=er+zero
      else   
!******* fill in Jij ***********
jij=zero
if(pp==0)then
   jij(ll,ll)=one
else
jij(pp,pp)=one
jij(ll,ll)=one
jij(pp,ll)=-one
jij(ll,pp)=-one
jije=jij
end if
jijb=zero

!*** form product akli*jij*akli and store in jij ********
ism1=matmul(jij,akli)
jij=matmul(akli,ism1)

!****************************************
!do the kronecker product

do i=1,ne
   kk=(i-1)
   do j=1,ne
      zq=(j-1)
      do k=1,3
      jijb(3*kk+k,3*zq+k)=jij(i,j)
      end do
   end do
end do


!*************************************
!electron repulsion integral proper

iv3=zero
i3=zero
do i=1,skp
   do j=1,skp
      iv3(i)=iv3(i)+jijb(i,j)*ee(j)
   end do
end do
do i=1,skp
   i3=i3+ee(i)*iv3(i)
end do

i2=zero
do i=1,ne
   i2=i2+ism1(i,i)
end do

i1=i3/i2

er=ncv(pp+1)*ncv(ll+1)*two*ov*f0(i1)/dsqrt(i2*pi)
ert=ert+er
!write(*,*)'er,ert',er,ert


end if
end do
end do
!stop
zht=zero
do i=1,ne
	zht=zht+ncv(i+1)*st(3*i)
end do
zht=ov*zht*efs

END SUBROUTINE matel
